activerecord/lib/active_record/associations/collection_association.rb

```diff
@@ -85,7 +85,8 @@ def ids_writer(ids)
       end
 
       def reset
-        super
```

Comment:
The distinction is around the type of the `@target` variable. In singular associations, it can either be `nil` or an ActiveRecord instance. In collection associations, it should always be an array — `nil` is invalid. Here, the call to `super` initializes it to `nil`, with the subclass method immediately reinitializing it to `[]` and setting two more instance variables specific to the subclass. 

Perhaps it would be cleaner to have the class define a `default_target` method which could be implemented in a more direct way?

Reply:
> In singular associations, it can either be `nil` or an ActiveRecord instance. In collection associations, it should always be an array — `nil` is invalid.

It seems like `SingularAssociation` doesn't implement the `reset` method (but does inherit from `Association`). Maybe we should implement `reset` in `SingularAssociation` and have it set `@target` to nil? Then we can keep the `super` call here.